internal/planner/planner.go

```diff
@@ -1091,11 +1091,19 @@ func (p *Planner) planUnifyVar(a ast.Var, b *ast.Term, iter planiter) error {
 
 	return p.planTerm(b, func() error {
 		target := p.newLocal()
-		p.vars.Put(a, target)
-		p.appendStmt(&ir.AssignVarStmt{
-			Source: p.ltarget,
-			Target: target,
-		})
+		la, ok := p.vars.GetOp(a)
```

Comment:
If I understand the logic here correctly, this new chunk of code is for the fallthrough case where `p.vars.GetOp(a)` failed (on L1088-1090), so we try planning around `b`'s Term instead.

I'm not sure how the second `p.vars.GetOp(a)` would ever succeed if the first one failed, but maybe I'm missing something important? :thinking: If it never succeeds, then that means we're *always* getting the `ok == false` case.

Reply:
This is what happens, roughly:
1. we're planning `x = input.foo[x]`
2. the unification planning sees that lhs `x`, has it as unknown (L1088), and plans the unification with the rhs
3. it sees that rhs is a ref, with an unknown term `x`, so it'll do a scan,
4. in the scan, we're assigning the key to `x`,
5. coming back to unification, we're now assigning the result of planning the ref (i.e. its _value_) to `x`.

What this hadn't been accounting for is that `planTerm(b)` _lets us know `x`_ which we didn't when we've started planning the unification. When we look at `input.foo[x]`, we'll know that `x` is its _key_.

So the added inner `p.vars.GetOp(a)` succeeds because it's the _key_ of the scan done for the ref. We then have to assert equality on that and the value that we've gotten, `p.ltarget`, i.e. the value of `planTerm(b)`.

Reply:
Okay, that's the missing context I did not have there, that I had figured I was missing. :smiley: